### PR TITLE
Clean up exit behavior and error handling

### PR DESCRIPTION
**Pull request by awly**
Users can now choose between os.Exit on errors and handling errors themselves.
Returned error contains usage and an actual error from flag.Parse if that fails.

This is just conceptual approach, perhaps we can utilize FlagSet.Usage better.

### CODE REVIEW

File: flagconf.go [hunk 1]
```diff
@@ -166,6 +186,19 @@ func Parse(path string, config interface{}) error {
...
+// MustParse is like Parse except if parsing fails it prints the error and
+// exits the program.
+func MustParse(path string, config interface{}) {
+	if err := ParseStrings(os.Args, path, config, false); err != nil {
+		if ferr, ok := err.(FlagError); ok {
+			fmt.Fprintln(os.Stderr, ferr.Usage)
+		} else {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		os.Exit(2)
```
Comment (cespare) [inline flagconf.go:211]: Why 2?

Comment (awly) [inline flagconf.go:211]: Matches behavior of package flag:
http://golang.org/src/flag/flag.go?#L811

File: flagconf.go [hunk 2]
```diff
@@ -54,7 +56,25 @@ func ParseStrings(args []string, path string, config interface{}, allowNoConfigF
 	}
 
 	// Override any settings with configured flags
-	return flagset.Parse(args[1:])
+	if err = flagset.Parse(args[1:]); err != nil {
+		// In case flag parsing fails, return a custom error containing usage
+		// info if user wants to print it.
+		buf := &bytes.Buffer{}
+		flagset.SetOutput(buf)
```
Comment (cespare) [inline flagconf.go]: Because this is after `flagset.Parse()`, it won't capture the printing of the error. For instance, this code:

``` go
if err := flagconf.Parse("config.toml", &config); err != nil {
    fmt.Println(err)
}
```

makes this happen:

```
$ ./prog  -asdf
flag provided but not defined: -asdf
flag provided but not defined: -asdf
```

Is that intentional?

File: flagconf.go (hunk 1 above)
Comment (cespare) [inline flagconf.go:211]: OK thanks

**Comment by cespare** [comment]
Sorry for the delay on review. This mostly LGTM; just trying to think through the corner cases  (`-h`, `-badflag`).

I gave you commit bit; go ahead and merge when you're happy with this.

**Comment by awly** [comment]
No worries, thanks for the review.
Didn't quite test it well, thought that error message was coming from

```
fmt.Fprintln(os.Stderr, ferr.Usage)
```

after type assertion.

Will patch that up in a bit.

**Comment by awly** [comment]
@cespare fixed the `-badflag` issue locally.
What do you think is the correct handling of `-h`?
We can add a helper function `IsHelp(error)` so that user doesn't have to do 

```
err.(flagconf.FlagError).Err == flag.ErrHelp
```

or have `FlagError.Error()` append `Usage` to the string in that case.

**Comment by cespare** [comment]
@alytvynov Another issue is: if you call `MustParse` and then run `prog -h`, it will print just the usage out (but nothing else):

```
$ ./prog -h
  -addr="": listen address (with port)
  -maxprocs=4: maximum OS threads

```

including that extra newline at the end. Instead, I think it should print exactly the same as when you are using `flag.Parse()`.

**Comment by cespare** [comment]
@alytvynov `IsHelp` SGTM.

> or have `FlagError.Error()` append `Usage` to the string in that case.

Do you mean prepend?

**Comment by awly** [comment]
Ok, i think i got all of them.
Removed trailing newline from `FlagError.Usage` and pre-pended `Usage of <progname>:`.

This matches the output of package flag on tip exactly (r did change some things in flag's output formatting so it may be slightly different from 1.4.2).

IsHelp added too, although i'd expect most users to use MustParse.

PTAL

**Comment by cespare** [comment]
Thanks, LGTM. Merged.
